File: leetcode3/WordsWithinTwoEditsofDictionary/a.py
# -*- coding: utf-8 -*-
# Copyright (C) 2023 by iamslash

from typing import List

# 243ms 50.52% 14.1MB 48.96%
class Solution:
    def twoEditWords(self, queries: List[str], dictionary: List[str]) -> List[str]:
        ans = []
        for query in queries:
            for word in dictionary:
                if sum(c1 != c2 for c1, c2 in zip(query, word)) <= 2:
                    ans.append(query)
                    break
        return ans

if __name__ == '__main__':
    sln = Solution()
    print(sln.twoEditWords(["word","note","ants","wood"],
                           ["wood","joke","moat"]))

File: leetcode3/WordsWithinTwoEditsofDictionary/a.cpp
/* Copyright (C) 2023 by iamslash */

#include <cstdio>
#include <vector>
#include <string>

using namespace std;

// 39ms 41.46% 28.4MB 11.58%
// brute force
// O(QD) O(Q)
class Solution {
private:
  int dist(string a, string b) {
    int n = a.size(), cnt = 0;
    for (int i = 0; i < n; ++i) {
      if (a[i] != b[i] && ++cnt > 2) {
        break;
      }
    }
    return cnt;
  }
public:
  vector<string> twoEditWords(vector<string>& queries, vector<string>& dictionary) {
    vector<string> ans;
    for (string query : queries) {
      for (string word : dictionary) {
        if (dist(query, word) <= 2) {
          ans.push_back(query);
          break;
        }
      }
    }
    return ans;
  }
};

int main() {
  return 0;
}
